GetIntoTeachingApi/Services/CandidateAccessTokenService.cs

```diff
@@ -0,0 +1,60 @@
+﻿using GetIntoTeachingApi.Models;
+using OtpNet;
+using System;
+using System.Text;
+
+namespace GetIntoTeachingApi.Services
+{
+    public class CandidateAccessTokenService : ICandidateAccessTokenService
+    {
+        // The amount of time a user has to verify their access token is:
+        // (VerificationWindow * StepsInSeconds) + Remaining Seconds in Current Step
+        public static readonly int VerificationWindow = 2;
+        public static readonly int StepInSeconds = 30;
+        private static readonly int Length = 6;
+
+        public string GenerateToken(string email)
+        {
+           var  totp = CreateTotp(email);
+            return totp.ComputeTotp();
+        }
+
+        public bool IsValid(CandidateAccessTokenChallenge challenge)
+        {
+            return IsValid(challenge, DateTime.UtcNow);
+        }
+
+        public bool IsValid(CandidateAccessTokenChallenge challenge, DateTime timestamp)
+        {
+            if (!challenge.HasToken() || !challenge.SharedSecretMatches(SharedSecret()))
+            {
+                return false;
+            }
+
+            var totp = CreateTotp(challenge.Email);
+
+            long timeWindowUsed;
+            return totp.VerifyTotp(
+                timestamp,
+                challenge.Token,
+                out timeWindowUsed,
+                new VerificationWindow(previous: VerificationWindow, future: VerificationWindow)
+            );
+        }
+
+        private Totp CreateTotp(string email)
+        {
+            return new Totp(CompoundSharedSecretBytes(email), totpSize: Length, step: StepInSeconds);
+        }
+
+        private byte[] CompoundSharedSecretBytes(string email)
+        {
+            return Encoding.ASCII.GetBytes(email + SharedSecret());

```

Comment:
You mentioned that this value gets truncated, do you know if its hashed prior to truncation - or is it the resultant string thats truncated?

If its the compound string I'd suggest hashing this. Either SHA1 or SHA2-256, probably the latter if its going to get truncated - I don't think the cost of hashing will be a consideration.

```
return Encoding.ASCII.GetBytes(HypotheticalSHA2Function(email + SharedSecret())) ;
```

Reply:
The key gets hashed with the current time prior to truncation - the function is:

```
TOTP(K,T) = Truncate(HMAC-SHA-1(K,T))
```